File: BennuLib/IO/DivFntSpriteAssetDecoder.vb
Option Infer On

Imports PkSprite.Core

Namespace Bennu.IO
    Public Class DivFntFpgDecoder
        Inherits NativeDecoder(Of SpriteAsset)

        Public Overrides ReadOnly Property MaxSupportedVersion As Integer = &H0

        Protected Overrides ReadOnly Property KnownFileExtensions As String() = {"fnt"}

        Protected Overrides ReadOnly Property KnownFileIds As String() = {"fnt"}

        Protected Overrides Function ReadNativeFormat(magic As Magic, reader As NativeFormatReader) As SpriteAsset

            Dim pal = Palette.Create(VGAtoColors(reader.ReadPalette))
            reader.ReadUnusedPaletteGamma()

            Dim fontInfo As Integer = reader.ReadInt32

            Dim characters(255) As GlyphInfo
            For n = 0 To 255
                characters(n) = reader.ReadGlyphInfo()
            Next

            Dim fpg As New SpriteAsset
            For Each character In characters
                Dim dataLength = character.Height * character.Width

                If character.FileOffset = 0 Or dataLength = 0 Then Continue For

                Dim graphicData = reader.ReadBytes(dataLength)

                Dim pixels = IndexedPixel.CreateBufferFromBytes(graphicData)
                Dim map = Sprite.Create(character.Width, character.Height, pixels)
                fpg.Add(fpg.FindFreeCode, map)
            Next

            Return fpg
        End Function
    End Class
End Namespace

File: BennuLib.Sharp/IO/DivFntSpriteAssetDecoder.cs
using static BennuLib.IO.NativeFormat;

namespace BennuLib.IO
{
	public class DivFntFpgDecoder : NativeDecoder<SpriteAsset>
	{

		public override int MaxSupportedVersion { get; }

		protected override string[] KnownFileExtensions { get; }

		protected override string[] KnownFileMagics { get; }

		protected override SpriteAsset ReadBody(Header header, NativeFormatReader reader)
		{

			var pal = Palette.Create(VGAtoColors(reader.ReadPalette()));
			reader.ReadUnusedPaletteGamma();

			int fontInfo = reader.ReadInt32();

			GlyphInfo[] characters = new GlyphInfo[256];
			for (var n = 0; n <= 255; n++) {
				characters[n] = reader.ReadGlyphInfo();
			}

			SpriteAsset fpg = new SpriteAsset();
			foreach (var character in characters) {
				var dataLength = character.Height * character.Width;

				if (character.FileOffset == 0 | dataLength == 0)
					continue;

				var graphicData = reader.ReadBytes(dataLength);

				var pixels = IndexedPixel.CreateBufferFromBytes(graphicData);
				var map = Sprite.Create(character.Width, character.Height, pixels);
				fpg.Add(fpg.FindFreeId(), ref map);
			}

			return fpg;
		}
	}
}
